crates/subspace-networking/src/behavior/persistent_parameters.rs

```diff
@@ -0,0 +1,248 @@
+    /// On object creation it starts a job for networking parameters cache handling.
+    pub fn new(path: &Path) -> Result<Self, NetworkParametersPersistenceError> {
+        let mut options = Options::with_columns(path, 1);
+        // We don't use stats
+        options.stats = false;
+
+        let db = Db::open_or_create(&options)?;
+        let column_id = 0u8;
+        let object_id = b"global_networking_parameters_key";
+
+        // load known peeers cache.
+        let cache = db
+            .get(column_id, object_id)?
+            .map(|data| {
+                let result = serde_json::from_slice::<NetworkingParametersDto>(&data).map(|data| {
+                    let cache: LruCache<PeerId, HashSet<Multiaddr>> = data.into();
+
+                    cache
+                });
+
+                if result.is_ok() {
+                    trace!("Networking parameters loaded from DB");
+                }
+
+                result
+            })
+            .unwrap_or_else(|| Ok(LruCache::new(PEER_CACHE_SIZE)))?;
+
+        Ok(Self {
+            db: Arc::new(db),
+            column_id,
+            object_id,
+            known_peers: cache,
+            networking_parameters_save_delay: Self::default_delay(),
+        })
+    }
+
+    // Helps create a copy of the internal LruCache
+    fn clone_known_peers(&self) -> LruCache<PeerId, HashSet<Multiaddr>> {
+        let mut known_peers = LruCache::new(self.known_peers.cap());
+
+        for (peer_id, addresses) in self.known_peers.iter() {
+            known_peers.push(*peer_id, addresses.clone());
+        }
+
+        known_peers
+    }
+
+    /// Creates a reference to the `NetworkingParametersRegistry` trait implementation.
+    pub fn boxed(self) -> Box<dyn NetworkingParametersRegistry> {
+        Box::new(self)
+    }
+
+    // Create default delay for networking parameters.
+    fn default_delay() -> Pin<Box<Fuse<Sleep>>> {
+        Box::pin(sleep(Duration::from_secs(DATA_FLUSH_DURATION_SECS)).fuse())
+    }
+}
+
+#[async_trait]
+impl NetworkingParametersRegistry for NetworkingParametersManager {
+    async fn add_known_peer(&mut self, peer_id: PeerId, addresses: Vec<Multiaddr>) {
+        let addr_set = addresses.iter().cloned().collect::<HashSet<_>>();
+
+        if let Some(addresses) = self.known_peers.get_mut(&peer_id) {
+            *addresses = addresses.union(&addr_set).cloned().collect()
+        } else {
+            self.known_peers.push(peer_id, addr_set);
+        }
+    }
+
+    async fn known_addresses(&self, peer_number: usize) -> Vec<(PeerId, Multiaddr)> {
+        self.known_peers
+            .iter()
+            .take(peer_number)
+            .flat_map(|(peer_id, addresses)| addresses.iter().map(|addr| (*peer_id, addr.clone())))
+            .map(|(peer_id, addr)| {
+                // remove p2p-protocol suffix if any
+                let mut modified_address = addr.clone();
+
+                if let Some(Protocol::P2p(_)) = modified_address.pop() {
```

Comment:
This seems unnecessary, database isn't storing addresses with `p2p` at the end anyway.

Reply:
I'm not following here, why not? We get addresses with `p2p` protocol from the `ConnectedPoint::Dialer`.

Reply:
Then we need to filter them during insertion and store in canonical form without `p2p` protocol at the end (it is already in the key). Makes more sense to me that way instead of checking this every time on retrieval.